lib/src/main/java/graphql/nadel/validation/NadelHydrationArgumentValidation.kt

```diff
@@ -0,0 +1,181 @@
+            if (targetType.name == Scalars.GraphQLID.name &&
+                    (typeToAssign.name == Scalars.GraphQLString.name || typeToAssign.name == Scalars.GraphQLInt.name)) {
+                return true
+            }
+            if (targetType is GraphQLInterfaceType && typeToAssign is GraphQLImplementingType) {
+                return typeToAssign.interfaces.contains(targetType)
+            }
+            return false
+        }
+
+        private fun validateListArg(
+                hydrationSourceFieldType: GraphQLType,
+                actorFieldArgType: GraphQLType,
+                parent: NadelServiceSchemaElement,
+                overallField: GraphQLFieldDefinition,
+                remoteArg: RemoteArgumentDefinition,
+                hydration: UnderlyingServiceHydration
+        ): NadelSchemaValidationError? {
+            var hydrationSourceFieldInnerType: GraphQLType = hydrationSourceFieldType.unwrapNonNull().unwrapOne()
+            var actorFieldArgInnerType: GraphQLType = actorFieldArgType.unwrapNonNull().unwrapOne()
+            val errorExists = getHydrationInputErrors(
+                    hydrationSourceFieldInnerType,
+                    actorFieldArgInnerType,
+                    parent,
+                    overallField,
+                    remoteArg,
+                    hydration
+            ) != null
+            if (errorExists) {
+                return NadelSchemaValidationError.IncompatibleHydrationArgumentType(
+                        parent,
+                        overallField,
+                        remoteArg,
+                        hydrationSourceFieldType,
+                        actorFieldArgType,
+                )
+            }
+            return null
+        }
+
+        private fun validateInputObjectArg(
+                hydrationSourceFieldType: GraphQLType,
+                actorFieldArgType: GraphQLType,
+                parent: NadelServiceSchemaElement,
+                overallField: GraphQLFieldDefinition,
+                remoteArg: RemoteArgumentDefinition,
+                hydration: UnderlyingServiceHydration
+        ): NadelSchemaValidationError? {
+            val unwrappedHydrationSourceFieldType = hydrationSourceFieldType.unwrapNonNull() as GraphQLObjectType
+            val unwrappedActorFieldArgType = actorFieldArgType.unwrapNonNull() as GraphQLInputObjectType
+
+            val hydrationInnerFields: Map<String, GraphQLType> = unwrappedHydrationSourceFieldType.fields.associate { it.name to it.type as GraphQLType }
+            val actorInnerFields: Map<String, GraphQLType> = unwrappedActorFieldArgType.fields.associate { it.name to it.type as GraphQLType }
+
+            for (innerField in actorInnerFields){
+                val actorInnerFieldName = innerField.key
+                val actorInnerFieldType = innerField.value
+                val hydrationType = hydrationInnerFields.get(actorInnerFieldName)
+                if (hydrationType == null) {
```

Comment:
Mentioned offline that we need to check if the input field is required or not.

Reply:
done, thanks